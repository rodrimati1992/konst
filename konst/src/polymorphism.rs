//! Miscelaneous items used for emulating polymorphism without trait methods.
//!
//! # `typewit`
//!
//! This crate uses [`typewit`] for emulating trait based polymorphism in const fns,
//! because trait methods cannot be called in `const fn`s on stable(as of Rust 1.69.0)
//!

/// Markers used to classify types, used as `Kind` associated types.
pub mod kinds {
    #[doc(inline)]
    pub use konst_kernel::polymorphism::kinds::*;
}

#[doc(no_inline)]
pub use typewit::{
    self, simple_type_witness, type_fn, CallFn, HasTypeWitness, MakeTypeWitness, TypeEq, TypeFn,
    TypeWitnessTypeArg,
};

include! {"polymorphism/type_eq_projection_fn.rs"}
