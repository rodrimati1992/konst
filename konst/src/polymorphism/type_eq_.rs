
/// Gets a constant of `Te`, 
/// which is (by convention) an enum that allows type-based dispatch 
/// through the use of [`TypeEq`].
/// 
/// # Example 
/// 
/// This example shows how you can make a `const fn` that converts both 
/// `&str` and `&[u8]` to `&str`
/// 
/// ```rust
/// use konst::polymorphism::{InTypeEqEnum, TypeEq};
/// 
/// const fn str_try_from<'a, T, const L: usize>(input: T) -> Result<&'a str, std::str::Utf8Error>
/// where
///     T: Copy + InTypeEqEnum<StrTryFrom<'a, T, L>>
/// {
///     match T::TEQ_ENUM {
///         StrTryFrom::Str(teq) => Ok(teq.to_right(input)),
///         StrTryFrom::Bytes(teq) => {
///             let bytes = teq.to_right(input);
///             std::str::from_utf8(bytes)
///         }
///         StrTryFrom::Array(teq) => {
///             // this requires care not to infinitely recurse
///             let slice: &[_] = teq.to_right(input);
///             str_try_from(slice)
///         }
///     }
/// }
/// 
/// assert_eq!(str_try_from("hello"), Ok("hello"));
/// 
/// assert_eq!(str_try_from(&[b'w', b'o', b'r', b'l', b'd']), Ok("world"));
/// 
/// assert_eq!(str_try_from(b"foo bar" as &[_]), Ok("foo bar"));
/// 
/// 
/// impl<'a> InTypeEqEnum<StrTryFrom<'a, Self, 0>> for &'a str {
///     const TEQ_ENUM: StrTryFrom<'a, Self, 0> = StrTryFrom::Str(TypeEq::NEW);
/// }
/// 
/// impl<'a> InTypeEqEnum<StrTryFrom<'a, Self, 0>> for &'a [u8] {
///     const TEQ_ENUM: StrTryFrom<'a, Self, 0> = StrTryFrom::Bytes(TypeEq::NEW);
/// }
/// 
/// impl<'a, const L: usize> InTypeEqEnum<StrTryFrom<'a, Self, L>> for &'a [u8; L] {
///     const TEQ_ENUM: StrTryFrom<'a, Self, L> = StrTryFrom::Array(TypeEq::NEW);
/// }
/// 
/// // `#[non_exhausitve]` allows adding more supported types to the set.
/// #[non_exhaustive]
/// pub enum StrTryFrom<'a, T, const L: usize> {
///     Str(TypeEq<T, &'a str>),
///     Bytes(TypeEq<T, &'a [u8]>),
///     Array(TypeEq<T, &'a [u8; L]>),
/// }
/// 
/// ```
pub use konst_kernel::type_eq::InTypeEqEnum;


/// Value-level proof that `L` is the same type as `R`
///
/// This type can be used to prove that `L` and `R` are the same type,
/// because it can only be constructed with `TypeEq::<L, L>::NEW`,
/// where both type arguments are the same type.
#[doc(inline)]
pub use konst_kernel::type_eq::type_eq::TypeEq;