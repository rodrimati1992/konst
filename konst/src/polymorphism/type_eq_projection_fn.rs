/// Declares a function for converting a `TypeEq<L, R>`
/// to `TypeEq<Foo<L>, Foo<R>>`.
/// 
/// As an alternative to this macro, you can look at [`TypeEq::project`] and [`TypeEq::map`]
/// 
/// [**examples below**](#examples)
/// 
/// [**syntax example**](#syntax)
/// 
/// # Limitations
/// 
/// This macro has the following limitations:
/// - It only accepts module paths for a type,
/// followed by the generic parameters of that type,
/// no concrete generic arguments are allowed.
/// 
/// - It can only map one type parameter, the `T` parameter.
/// 
/// - It cannot parse trait bounds in the type parameter list written 
/// the normal way, they must be wrapped in parentheses.
/// 
/// - The `T` type parameter can only be bounded in the parameter list
/// 
/// - The `T` type parameter cannot appear in any trait bounds.
/// 
/// The first two limitations can be worked around by passing a type alias
/// to the macro.
///
/// # Examples
/// 
/// ### Basic
/// 
/// This example shows what the macro does,
/// the [motivating example](#motivating-example) shows why one would use it.
/// 
/// ```rust
/// use konst::polymorphism::{TypeEq,  type_eq_projection_fn};
/// 
/// #[derive(Debug, PartialEq)]
/// struct Foo<T, const N: usize>([T; N]);
/// 
/// // This macro invocation generates:
/// // const fn project_to_foo<L, R, const N: usize>(
/// //     _: TypeEq<L, R>,
/// // ) -> TypeEq<Foo<L, N>, Foo<R, N>>
/// type_eq_projection_fn!{
///     // `T` must be both the function parameter, and in the return type.
///     const fn project_to_foo(T) -> Foo<T, const N: usize>
/// }
/// 
/// // a toy example to demonstrate what projecting a TypeEq does
/// const fn get_foo<'a, R>(te: TypeEq<&'a str, R>) -> Foo<R, 2> {
///     // The type annotation is for the reader
///     let te: TypeEq<Foo<&'a str, 2>, Foo<R, 2>> =
///         project_to_foo::<&'a str, R, 2>(te);
/// 
///     te.to_right(Foo(["foo", "bar"]))
/// }
/// 
/// assert_eq!(get_foo(TypeEq::NEW), Foo(["foo", "bar"]));
/// 
/// ```
/// 
/// ### Motivating example
/// 
/// ```rust
/// use konst::polymorphism::{
///     HasTypeWitness,
///     MakeTypeWitness,
///     TypeEq, 
///     TypeWitnessTypeArg,
///     type_eq_projection_fn,
/// };
/// 
/// fn main() {
///     assert_eq!(Foo(3, false).transform(), Foo(13, false));
///     assert_eq!(Foo("hello", "world").transform(), Foo("mapped", "world"));
/// }
/// 
/// #[derive(Debug, PartialEq)]
/// struct Foo<T, U: Copy>(T, U);
/// 
/// // This macro invocation generates:
/// // const fn project_to_foo<L, R, U>(
/// //     _: TypeEq<L, R>,
/// // ) -> TypeEq<Foo<L, U>, Foo<R, U>>
/// type_eq_projection_fn!{
///     // The `Copy` bound needs to be wrapped in parentheses in `U: (Copy)` to
///     // simplify parsing of trait bounds in the generic parameter list.
///     // 
///     // note: trait bounds are written normally in where clauses,
///     //       they must be unparenthesized.
///     const fn project_to_foo(T) -> Foo<T, U: (Copy)>
/// }
/// 
/// impl<T, U: Copy> Foo<T, U> {
///     const fn transform<'a>(self) -> Foo<T, U>
///     where
///         T: Copy + HasTypeWitness<TheWitness<'a, T>>,
///     {
///         match T::WITNESS {
///             TheWitness::U8(te) => {
///                 // the type annotation is just for the reader
///                 let te: TypeEq<Foo<T, U>, Foo<u8, U>> = project_to_foo(te);
///                 let bar: Foo<u8, U> = te.to_right(self);
///
///                 te.to_left(Foo(bar.0 + 10, bar.1))
///             }
///             TheWitness::Str(te) => {
///                 // the type annotation is just for the reader
///                 let te: TypeEq<Foo<T, U>, Foo<&str, U>> = project_to_foo(te);
///                 te.to_left(Foo("mapped", self.1))
///             }
///         }
///     }
/// }
/// 
/// // A type witness, a pattern documented in `konst::docs::type_witnesses`
/// //
/// // Simply put, type witnesses emulate matching over a range of types
/// // (not values of those types, the types themselves).
/// enum TheWitness<'a, T> {
///     U8(TypeEq<T, u8>),
///     Str(TypeEq<T, &'a str>),
/// }
/// 
/// impl<T> TypeWitnessTypeArg for TheWitness<'_, T> {
///     type Arg = T;
/// }
/// 
/// impl MakeTypeWitness for TheWitness<'_, u8> {
///     const MAKE: Self = Self::U8(TypeEq::NEW);
/// }
/// 
/// impl<'a> MakeTypeWitness for TheWitness<'a, &'a str> {
///     const MAKE: Self = Self::Str(TypeEq::NEW);
/// }
/// 
/// ```
/// 
/// ### Syntax
/// 
/// This example demonstrates all the syntax that this macro supports.
/// 
/// ```rust
/// # use std::fmt::Debug;
/// # use konst::polymorphism::type_eq_projection_fn;
/// #
/// # extern crate self as foo;
/// #
/// # #[derive(Debug, PartialEq, Clone)]
/// # pub struct Ty<'a, 'b: 'a, U: 'a + Debug, const N: usize>(&'a &'b [U; N]);
/// #
/// // This macro invocation generates this function:
/// // 
/// // pub const fn project<'a, 'b, L, R, const N: usize>(
/// //     _: TypeEq<L, R>
/// // ) -> TypeEq<::foo::Ty<'a, 'b, L, N>, ::foo::Ty<'a, 'b, R, N>>
/// // where
/// //     'b: 'a,
/// //     L: 'b + Debug,
/// //     R: 'b + Debug,
/// //     [u32; N]: 'a + core::fmt::Debug
/// type_eq_projection_fn!{
///     /// Documentation for the generated function
///     // 
///     // Without the `const` qualifier, the generated function is non-`const`.
///     // 
///     // `T` must be both the function parameter, and in the return type.
///     pub const fn project(T) -> ::foo::Ty<
///         'a,
///         'b: 'a,
///         // trait bounds in the type parameter list must be parenthesized
///         T: ('b +  Debug), 
///         const N: usize,
///     >
///     where
///         // trait bounds in the where clause are unparenthesized
///         [u32; N]: 'a + core::fmt::Debug,
/// }
/// # fn main(){}
/// ```
/// 
pub use konst_kernel::type_eq_projection_fn;