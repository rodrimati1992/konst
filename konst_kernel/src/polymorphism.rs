pub mod kinds {

    /// Marker for user-defined types that can be converted into const iterators
    pub enum IsIntoIterKind {}

    /// Marker for const iterators
    pub enum IsIteratorKind {}

    /// Marker for references.
    pub enum IsRefKind {}

    /// Marker for non-standard library types.
    pub enum IsNotStdKind {}

    /// Marker for standard library types.
    pub enum IsStdKind {}
}
