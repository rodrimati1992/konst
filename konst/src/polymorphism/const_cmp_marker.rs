//! Marker trait for types that implement the const formatting methods.
//!
//!

use crate::polymorphism::CmpWrapper;

use core::marker::PhantomData;

////////////////////////////////////////////////////////////////////////////////

/// Marker trait for types that implement the const comparison methods.
///
/// # Implementors
///
/// Types that implement this trait are also expected to implement at least one of
/// these inherent methods:
///
/// ```ignore
/// // use std::cmp::Ordering;
///
/// const fn const_eq(&self, other: &Self) -> bool
///
/// const fn const_cmp(&self, other: &Self) -> Ordering
///
/// ```
///
/// # Coercions
///
/// The [`Kind`](#associatedtype.Kind) associated type
/// is used in the [`IsAConstCmpMarker`] marker type
/// to automatically wrap types in [`CmpWrapper`] if they're from the standard library,
/// otherwise leaving them unwrapped.
///
///
/// # Example
///
/// ### Manual Implementation
///
/// ```
/// use konst::{
///     polymorphism::{ConstCmpMarker, IsNotStdKind},
///     const_cmp, const_eq, try_equal,
/// };
///
/// use std::cmp::Ordering;
///
///
/// struct MyType {
///     x: &'static str,
///     y: &'static [u16],
/// }
///
/// impl ConstCmpMarker for MyType {
///     // These are the only associated types you're expected to use in
///     // impls for custom types.
///     type Kind = IsNotStdKind;
/// }
///
/// impl MyType {
///     pub const fn const_eq(&self, other: &Self) -> bool {
///         const_eq!(self.x, other.x) &&
///         const_eq!(self.y, other.y)
///     }
///
///     pub const fn const_cmp(&self, other: &Self) -> Ordering {
///         try_equal!(const_cmp!(self.x, other.x));
///         try_equal!(const_cmp!(self.y, other.y))
///     }
/// }
///
/// const CMPS: [(Ordering, bool); 4] = {
///     let foo = MyType{x: "hello", y: &[3, 5, 8, 13]};
///     let bar = MyType{x: "world", y: &[3, 5, 8, 13]};
///
///     [
///         (const_cmp!(foo, foo), const_eq!(foo, foo)),
///         (const_cmp!(foo, bar), const_eq!(foo, bar)),
///         (const_cmp!(bar, foo), const_eq!(bar, foo)),
///         (const_cmp!(bar, bar), const_eq!(bar, bar)),
///     ]
/// };
///
/// assert_eq!(
///     CMPS,
///     [
///         (Ordering::Equal, true),
///         (Ordering::Less, false),
///         (Ordering::Greater, false),
///         (Ordering::Equal, true),
///     ]
/// );
///
///
///
/// ```
///
///
/// ### `Ã¬mpl_cmp`-based Implementation
///
/// You can use [`impl_cmp`] to implement this trait,
/// as well as define the same methods for
/// multiple implementations with different type arguments.
///
/// ```
/// use konst::{const_cmp, const_eq, impl_cmp, try_equal};
///
/// use std::cmp::Ordering;
///
///
/// struct MyType<'a, T> {
///     x: &'a str,
///     y: &'a [T],
/// }
///
/// impl_cmp!{
///     // The comparison functions are only implemented for these types.
///     impl['a] MyType<'a, bool>;
///     impl['a] MyType<'a, u16>;
///     impl['a] MyType<'a, &'static str>;
///
///     pub const fn const_eq(&self, other: &Self) -> bool {
///         const_eq!(self.x, other.x) &&
///         const_eq!(self.y, other.y)
///     }
///
///     pub const fn const_cmp(&self, other: &Self) -> Ordering {
///         try_equal!(const_cmp!(self.x, other.x));
///         try_equal!(const_cmp!(self.y, other.y))
///     }
/// }
///
/// const CMPS: [(Ordering, bool); 4] = {
///     let foo = MyType{x: "hello", y: &[3u16, 5, 8, 13]};
///     let bar = MyType{x: "world", y: &[3, 5, 8, 13]};
///
///     [
///         (const_cmp!(foo, foo), const_eq!(foo, foo)),
///         (const_cmp!(foo, bar), const_eq!(foo, bar)),
///         (const_cmp!(bar, foo), const_eq!(bar, foo)),
///         (const_cmp!(bar, bar), const_eq!(bar, bar)),
///     ]
/// };
///
/// assert_eq!(
///     CMPS,
///     [
///         (Ordering::Equal, true),
///         (Ordering::Less, false),
///         (Ordering::Greater, false),
///         (Ordering::Equal, true),
///     ]
/// );
///
/// ```
///
/// [`CmpWrapper`]: struct.CmpWrapper.html
/// [`impl_cmp`]: ../macro.impl_cmp.html
pub trait ConstCmpMarker {
    /// What kind of type this is, this can be one of:
    ///
    /// - [`IsStdKind`]: A standard library type.
    ///
    /// - [`IsRefKind`]: A reference type.
    ///
    /// - [`IsNotStdKind`]: A type that is not from the standard library.
    ///
    type Kind;
}

/// Marker type for standard library types.
pub struct IsStdKind;

/// Marker type for references.
pub struct IsRefKind;

/// Marker type for non-standard library types.
pub struct IsNotStdKind;

///////////////////////////////////////////////////////////////////////////////

impl<T, const N: usize> ConstCmpMarker for [T; N] {
    type Kind = IsStdKind;
}

impl<T> ConstCmpMarker for [T] {
    type Kind = IsStdKind;
}

impl ConstCmpMarker for str {
    type Kind = IsStdKind;
}

impl<T> ConstCmpMarker for &T
where
    T: ?Sized + ConstCmpMarker,
{
    type Kind = IsRefKind;
}

impl<T> ConstCmpMarker for &mut T
where
    T: ?Sized + ConstCmpMarker,
{
    type Kind = IsRefKind;
}

///////////////////////////////////////////////////////////////////////////////

/// A helper trait of [`ConstCmpMarker`], used for dereferencing.
pub trait ConstCmpUnref: ConstCmpMarker {
    /// What type `Self` becomes after removing all layers of references.
    type This: ?Sized + ConstCmpMarker;
}

impl<T> ConstCmpUnref for T
where
    T: ?Sized + ConstCmpMarker,
    T: ConstCmpUnrefHelper<<T as ConstCmpMarker>::Kind>,
    T::This_: ConstCmpMarker,
{
    type This = T::This_;
}

/// An implementation detail of [`ConstCmpUnref`].
pub trait ConstCmpUnrefHelper<Kind> {
    type This_: ?Sized;
}

impl<T: ?Sized> ConstCmpUnrefHelper<IsStdKind> for T {
    type This_ = T;
}

impl<T: ?Sized> ConstCmpUnrefHelper<IsNotStdKind> for T {
    type This_ = T;
}

impl<T: ?Sized + ConstCmpUnref> ConstCmpUnrefHelper<IsRefKind> for &T {
    type This_ = T::This;
}
impl<T: ?Sized + ConstCmpUnref> ConstCmpUnrefHelper<IsRefKind> for &mut T {
    type This_ = T::This;
}

///////////////////////////////////////////////////////////////////////////////

/// Hack used to automatically wrap standard library types inside [`CmpWrapper`],
/// while leaving user defined types unwrapped.
///
/// This can be constructed with he [`NEW` associated constant](#associatedconstant.NEW)
///
/// # Type parameters
///
/// `K` is `<T as ConstCmpMarker>::Kind`
/// The kind of type that `T` is,
/// [std types](./struct.IsStdKind.html),
/// [non-std types](./struct.IsNotStdKind.html).
///
/// `T` is `<R as ConstCmpUnref>::This`,
/// the `R` type after removing all layers of references.
///
/// `R`: Is a type that implements [`ConstCmpMarker`]
///
#[allow(clippy::type_complexity)]
pub struct IsAConstCmpMarker<K, T: ?Sized, R: ?Sized>(
    PhantomData<(
        PhantomData<fn() -> PhantomData<K>>,
        PhantomData<fn() -> PhantomData<T>>,
        PhantomData<fn() -> PhantomData<R>>,
    )>,
);

impl<K, T: ?Sized, R: ?Sized> Copy for IsAConstCmpMarker<K, T, R> {}

impl<K, T: ?Sized, R: ?Sized> Clone for IsAConstCmpMarker<K, T, R> {
    fn clone(&self) -> Self {
        *self
    }
}

impl<R, T> IsAConstCmpMarker<T::Kind, T, R>
where
    R: ?Sized + ConstCmpUnref<This = T>,
    T: ?Sized + ConstCmpMarker,
{
    /// Constructs an `IsAConstCmpMarker`
    pub const NEW: Self = Self(PhantomData);
}

impl<K, T: ?Sized, R: ?Sized> IsAConstCmpMarker<K, T, R> {
    /// Infers the type parameters by taking a reference to `R` .
    ///
    /// The `K` and `T` type parameters are determined by `R` in
    /// the [`NEW`] associated constant.
    ///
    /// [`NEW`]: #associatedconstant.NEW
    #[inline(always)]
    pub const fn infer_type(self, _: &R) -> Self {
        self
    }

    /// Removes layers of references by coercing the argument.
    #[inline(always)]
    pub const fn unreference(self, r: &T) -> &T {
        r
    }
}

/////////////////////////////////////////////////////////////////////////////

impl<T: ?Sized, R: ?Sized> IsAConstCmpMarker<IsNotStdKind, T, R> {
    /// An identity function, just takes `reference` and returns it.
    #[inline(always)]
    pub const fn coerce(self, reference: &T) -> &T {
        reference
    }
}

/////////////////////////////////////////////////////////////////////////////

impl<R: ?Sized> IsAConstCmpMarker<IsStdKind, str, R> {
    /// Wraps `reference` in a `CmpWrapper`.
    #[inline(always)]
    pub const fn coerce(self, reference: &str) -> CmpWrapper<&str> {
        CmpWrapper(reference)
    }
}

impl<T, R: ?Sized> IsAConstCmpMarker<IsStdKind, [T], R> {
    /// Wraps `reference` in a `CmpWrapper`.
    #[inline(always)]
    pub const fn coerce(self, reference: &[T]) -> CmpWrapper<&[T]> {
        CmpWrapper(reference)
    }
}

impl<T, R, const N: usize> IsAConstCmpMarker<IsStdKind, [T; N], R> {
    /// Wraps `reference` in a `CmpWrapper`.
    #[inline(always)]
    pub const fn coerce(self, reference: &[T; N]) -> CmpWrapper<&[T]> {
        CmpWrapper(reference)
    }
}
