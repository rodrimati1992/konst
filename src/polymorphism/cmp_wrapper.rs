use crate::polymorphism::{ConstCmpMarker, IsAConstCmpMarker, IsNotStdKind, IsStdKind};

pub struct CmpWrapper<T>(pub T);

impl<'a, T> CmpWrapper<&'a [T]> {
    /// For constructing from a reference to an array.
    ///
    /// With slices you can do `CmpWrapper(slice)` as well.
    #[inline(always)]
    pub const fn slice(x: &'a [T]) -> Self {
        Self { 0: x }
    }
}

impl<P> ConstCmpMarker for CmpWrapper<P> {
    type Kind = IsNotStdKind;
    type This = Self;
}

macro_rules! std_kind_impls {
    ($($ty:ty),* $(,)* ) => (
        $(
            impl ConstCmpMarker for $ty {
                type Kind = IsStdKind;
                type This = Self;
            }

            impl<T> IsAConstCmpMarker<IsStdKind, $ty, T> {
                /// Copies the value from `reference`, and wraps it in a `CmpWrapper`
                #[inline(always)]
                pub const fn coerce(self, reference: &$ty) -> CmpWrapper<$ty> {
                    CmpWrapper(*reference)
                }
            }

            impl CmpWrapper<$ty> {
                #[inline(always)]
                pub const fn const_eq(self, other: &$ty) -> bool {
                    self.0 == *other
                }
            }
        )*
    )
}

std_kind_impls! {
    i8, u8,
    i16, u16,
    i32, u32,
    i64, u64,
    i128, u128,
    isize, usize,
    bool, char,
}
