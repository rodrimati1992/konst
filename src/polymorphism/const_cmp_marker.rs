//! Marker trait for types that implement the const formatting methods.
//!
//!

use crate::polymorphism::CmpWrapper;

use core::marker::PhantomData;

////////////////////////////////////////////////////////////////////////////////

/// Marker trait for types that implement the const comparison methods.
///
pub trait ConstCmpMarker {
    /// What kind of type this is, this can be one of:
    ///
    /// - [`IsArrayKind`]: For slices, and arrays.
    ///
    /// - [`IsStdKind`]: Any other standard library type.
    ///
    /// - [`IsNotStdKind`]: Any type that is not from the standard library.
    ///
    /// [`IsArrayKind`]: ./struct.IsArrayKind.html
    /// [`IsStdKind`]: ./struct.IsStdKind.html
    /// [`IsNotStdKind`]: ./struct.IsNotStdKind.html
    type Kind;

    /// The type after dereferencing,
    /// implemented as `type This = Self;` for all non-reference types
    type This: ?Sized;
}

/// Marker type for arrays and slices,
/// used as the [`Kind`] associated type  in [`ConstCmpMarker`].
///
/// [`Kind`]: ./trait.ConstCmpMarker.html#associatedtype.Kind
/// [`ConstCmpMarker`]: ./trait.ConstCmpMarker.html
///
pub struct IsArrayKind<T>(PhantomData<T>);

/// Marker type for the remaining standard library types,,
/// used as the [`Kind`] associated type  in [`ConstCmpMarker`].
///
/// [`Kind`]: ./trait.ConstCmpMarker.html#associatedtype.Kind
/// [`ConstCmpMarker`]: ./trait.ConstCmpMarker.html
///
pub struct IsStdKind;

/// Marker type for non-standard library types,
/// used as the [`Kind`] associated type  in [`ConstCmpMarker`].
///
/// [`Kind`]: ./trait.ConstCmpMarker.html#associatedtype.Kind
/// [`ConstCmpMarker`]: ./trait.ConstCmpMarker.html
///
pub struct IsNotStdKind;

///////////////////////////////////////////////////////////////////////////////

/// Hack used to automatically wrap standard library types inside [`CmpWrapper`],
/// while leaving user defined types unwrapped.
///
/// # Type parameters
///
/// `K` is `<R as ConstCmpMarker>::Kind`
/// The kind of type that `T` is,
/// [a slice](./struct.IsArrayKind.html),
/// [other std types](./struct.IsStdKind.html),
/// [non-std types](./struct.IsNotStdKind.html).
///
/// `T` is `<R as ConstCmpMarker>::This`,
/// the `R` type after removing all layers of references.
///
/// # Coerce Method
///
/// The `coerce` method is what does the conversion from a `&T` depending on
/// the `K` type parameter:
///
/// - [`IsArrayKind`]: the reference is coerced to a slice, and wrapped in a [`CmpWrapper`].
///
/// - [`IsStdKind`]: the referenced value is copied, and wrapped in a [`CmpWrapper`].
///
/// - [`IsNotStdKind`]: the reference is simply returned as a `&T`.
///
/// [`IsArrayKind`]: ./struct.IsArrayKind.html
/// [`IsStdKind`]: ./struct.IsStdKind.html
/// [`IsNotStdKind`]: ./struct.IsNotStdKind.html
///
/// [`CmpWrapper`]: ../struct.CmpWrapper.html
///
#[allow(clippy::type_complexity)]
pub struct IsAConstCmpMarker<K, T: ?Sized, R: ?Sized>(
    PhantomData<(
        PhantomData<fn() -> PhantomData<K>>,
        PhantomData<fn() -> PhantomData<T>>,
        PhantomData<fn() -> PhantomData<R>>,
    )>,
);

impl<K, T: ?Sized, R: ?Sized> Copy for IsAConstCmpMarker<K, T, R> {}

impl<K, T: ?Sized, R: ?Sized> Clone for IsAConstCmpMarker<K, T, R> {
    fn clone(&self) -> Self {
        *self
    }
}

impl<R> IsAConstCmpMarker<R::Kind, R::This, R>
where
    R: ?Sized + ConstCmpMarker,
{
    /// Constructs an `IsAConstCmpMarker`
    pub const NEW: Self = Self(PhantomData);
}

impl<K, T: ?Sized, R: ?Sized> IsAConstCmpMarker<K, T, R> {
    /// Infers the type parameters by taking a reference to `R` .
    ///
    /// The `K` and `T` type parameters are determined by `R` in
    /// the [`NEW`] associated constant.
    ///
    /// [`NEW`]: #associatedconstant.NEW
    #[inline(always)]
    pub const fn infer_type(self, _: &R) -> Self {
        self
    }

    /// Removes layers of references by coercing the argument.
    #[inline(always)]
    pub const fn unreference(self, r: &T) -> &T {
        r
    }
}

/////////////////////////////////////////////////////////////////////////////

impl<U, T: ?Sized, R: ?Sized> IsAConstCmpMarker<IsArrayKind<U>, T, R> {
    /// Coerces an array to a slice, then wraps the slice in a `CmpWrapper`
    #[inline(always)]
    pub const fn coerce(self, slice: &[U]) -> CmpWrapper<&[U]> {
        CmpWrapper(slice)
    }
}

impl<T: ?Sized, R: ?Sized> IsAConstCmpMarker<IsNotStdKind, T, R> {
    /// An identity function, just takes `reference` and returns it.
    #[inline(always)]
    pub const fn coerce(self, reference: &T) -> &T {
        reference
    }
}

/////////////////////////////////////////////////////////////////////////////

#[cfg(feature = "const_generics")]
macro_rules! array_impls {
    ($($tt:tt)*) => {
        impl<T, const N: usize> ConstCmpMarker for [T; N] {
            type Kind = IsArrayKind<T>;
            type This = Self;
        }
    };
}

#[cfg(not(feature = "const_generics"))]
macro_rules! array_impls {
    ($($len:literal),* $(,)* ) => (
        $(
            impl<T> ConstCmpMarker for [T; $len] {
                type Kind = IsArrayKind<T>;
                type This = Self;
            }
        )*
    )
}

impl ConstCmpMarker for str {
    type Kind = IsStdKind;
    type This = Self;
}

impl<R: ?Sized> IsAConstCmpMarker<IsStdKind, str, R> {
    /// Wraps `reference` in a `CmpWrapper`.
    #[inline(always)]
    pub const fn coerce(self, reference: &str) -> CmpWrapper<&str> {
        CmpWrapper(reference)
    }
}

array_impls! {
    0,1,2,3,4,5,6,7,8,9,10,11,12,13,14,15,
    16,17,18,19,20,21,22,23,24,25,26,27,28,29,30,31,
    32,
}

impl<T> ConstCmpMarker for [T] {
    type Kind = IsArrayKind<T>;
    type This = [T];
}

impl<T> ConstCmpMarker for &T
where
    T: ?Sized + ConstCmpMarker,
{
    type Kind = T::Kind;
    type This = T::This;
}

impl<T> ConstCmpMarker for &mut T
where
    T: ?Sized + ConstCmpMarker,
{
    type Kind = T::Kind;
    type This = T::This;
}
