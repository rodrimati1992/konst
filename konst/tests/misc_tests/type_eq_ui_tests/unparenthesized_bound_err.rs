use konst::polymorphism::type_eq_projection_fn;

type_eq_projection_fn!{
    ///
    fn project_option(T) -> Option<T: core::fmt::Debug>
}

fn main(){}