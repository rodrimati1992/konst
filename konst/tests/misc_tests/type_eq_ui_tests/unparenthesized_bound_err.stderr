error: trait bounds in parameter list must be wrapped in parentheses, context: `T : core`
 --> tests/misc_tests/type_eq_ui_tests/unparenthesized_bound_err.rs:3:1
  |
3 | / type_eq_projection_fn!{
4 | |     ///
5 | |     fn project_option(T) -> Option<T: core::fmt::Debug>
6 | | }
  | |_^
  |
  = note: this error originates in the macro `$crate::__::__make_projection_parse_generics` which comes from the expansion of the macro `type_eq_projection_fn` (in Nightly builds, run with -Z macro-backtrace for more info)
